examples/qtmegachatapi/megachatapplication.cpp

```diff
@@ -0,0 +1,387 @@
+
+        gClient.reset();
+        remove((gAppDir+"/sid").c_str());
+        delete mainWin;
+        QMessageBox::critical(nullptr, tr("Logout"), tr("Your session has been closed remotely"));
+        createWindowAndClient();
+
+        gClient->loginSdkAndInit(nullptr)
+        .then([]()
+        {
+            KR_LOG_DEBUG("New client initialized with new session");
+            saveSid(gSdk->dumpSession());
+            QObject::connect(qApp, SIGNAL(lastWindowClosed()), &appDelegate, SLOT(onAppTerminate()));
+            gClient->connect(Presence::kInvalid);
+        })
+        .fail([](const promise::Error& err)
+        {
+            KR_LOG_ERROR("Error re-creating or logging in chat client after ESID: ", err.what());
+        });
+    }, NULL);*/
+}
+
+extern "C" void AppDelegate::myMegaPostMessageToGui(void* msg, void* appCtx, AppDelegate *appDelegate)
+{
+    QEvent* event = new GcmEvent(msg);
+    QApplication::postEvent(appDelegate, event);
+}
+
+
+MegaChatApplication::MegaChatApplication(std::string &dir)
+{
+         //Create Dir
+         //appDir = karere::createAppDir();
+
+         appDir=dir;
+
+         //Create Main Window
+         mainWin = new MainWindow();
+
+         //Create MegaAPI
+         megaApi=new ::mega::MegaApi("karere-native", appDir.c_str(), "Karere Native");
+         megaApi->setLogLevel(MegaApi::LOG_LEVEL_DEBUG);
+         megaApi->addListener(this);
+
+         //Create MegaChatAPI
+         megaChatApi=new ::megachat::MegaChatApi(megaApi);
+         megaChatApi->setLogLevel(MegaChatApi::LOG_LEVEL_DEBUG);
+         megaChatApi->addChatRequestListener(this);
+         megaChatApi->addChatListener(this);
+
+         #ifndef KARERE_DISABLE_WEBRTC
+            megaChatApi->addChatCallListener(mainWin);
+         #endif
+
+         mainWin->setMegaChatApi(megaChatApi);
+
+         //Create Login Dialog
+         loginDlg = new LoginDialog(nullptr);
+
+         //Create websocket
+         websocketsIO = new LibwsIO();
+}
+
+MegaChatApplication::~MegaChatApplication()
+{
+    megaApi->logout(nullptr);
+    megaChatApi->logout(nullptr);
+
+    marshallCall([this]()
+    {
+    delete megaChatApi;
+    delete megaApi;
+    qApp->quit();
+    delete websocketsIO;
+    delete mainWin;
+    loginDlg->destroy();
+    }, NULL);
+}
+
+int MegaChatApplication::init()
+{
+    return megaChatApi->init(sid);
+}
+
+void MegaChatApplication::login()
+{
+    auto pms = loginDlg->requestCredentials();
+    pms
+       .then([this](const std::pair<std::string, std::string>& cred)
+       {
+           assert(loginDlg);
+           loginDlg->setState(karere::IApp::ILoginDialog::kLoggingIn);
+           megaApi->login(cred.first.c_str(), cred.second.c_str());           
+       })
+       .fail([](const promise::Error& err)
+       {
+       });
+    QObject::connect(mainWin, SIGNAL(esidLogout()), &appDelegate, SLOT(onEsidLogout()));
```

Comment:
The logic for this signal should be re-implemented. The ESID stands for Expired Session ID, meaning that your current session is not valid anymore. It could happen if you close the session of this client from a different client, forcing a logout.
In this case, the behavior should be: 1. show a popup notifying the user about the event, and 2. show the login dialog again.